### DISCUSSION TITLE
Mojo::UserAgent logging request/response to file

### DISCUSSION THREAD

**Discussion by hussam-qasem** [mojolicious/mojo#2060] [Q&A] [answered]
I have a requirement to log all HTTP Requests / Responses to a file. How can I accomplish that using [`Mojo::UserAgent`](https://docs.mojolicious.org/Mojo/UserAgent)?

In SOAP::Lite, I do:
```perl
use SOAP::Lite 1.27 +trace => [ debug => \&Debug, ];
sub Debug { my ($in) = @_; log_to_file($in); }
```
Thank you!

**Comment by guest20** [comment]
You can do that kind of logging from a [prepare hook](https://metacpan.org/pod/Mojo::UserAgent#prepare ) when you set up your UserAgent: 

```perl
  has ua => sub {
    my $self = shift;
    use Mojo::UserAgent;
    use Mojo::Util qw( dumper );
    my $ua = Mojo::UserAgent->new;
    $ua->on(prepare => sub ($ua, $tx) { 
      $self->log->trace("about to do an http: ".dumper $tx);   
    });
    $ua
  }
```

you can also set the `User-Agent` header and all kinds of jazz from there so someone else with a badly written bot doesn't get your user agent banned too.

**Reply by hussam-qasem** [reply]
Thank you @guest20. Your solution works well for the Request, what about the HTTP Response?

side questions: why did you do `my $ua = Mojo::UserAgent->new;`? Couldn't you have used `$self->ua>on(prepare ...}); $self-ua` instead?

**Reply by guest20** [reply]
Because it's in the `has` callback, which returns the value for the attribute. Being there when the object is created makes sure nobody gets to call methods on the ua before you set your hooks up

**Comment by jberger** [comment]
I don't know if this will work for you, like if there are certain formatting requirements, but have you tried setting `MOJO_CLIENT_DEBUG=1` in your environment?

**Reply by hussam-qasem** [reply]
Thank you @jberger- I have not formatting requirements, and the output of `MOJO_CLIENT_DEBUG` works perfectly. Is it possible to redirect it to a file - or different files for different Models?

**Reply by jberger** [reply]
It is printed to stderr so you could redirect that to a file. It is a low-level directive however, it cannot be bound to specific models or even specific user agents. If needed you could see how it is implemented and copy it for your own needs.

**Reply by jberger** [reply, marked as answer]
Look in here for the DEBUG constant: https://github.com/mojolicious/mojo/blob/main/lib/Mojo/UserAgent.pm

Probably you just need to subscribe to the read and write events of the underlying socket like it does
